website_git.inc.php

```diff
@@ -111,6 +111,12 @@ function update($event_name, $data) {
 			//$this->_exec($gitBinary . ' --work-tree=' . escapeshellarg($docroot . "/web") . ' --git-dir=' . escapeshellarg($docroot . "/website.git") . ' config core.worktree "/web"');
 		}
 
+		if(!@is_dir($docroot.'/'.$docroot)){
+			$app->system->mkdirpath($docroot.'/'.$docroot, 0711, $username, $groupname);
```

Comment:
It is used to create a path for users in a jailkit chroot environment, as their root directory is the $docroot there is no `/var/www/clients/clientXX/webYY/` available to them, so that when the git hook tries to access `$docroot/website.git` it finds no directory.  
Creating the `$docroot . '/' . $docroot` path and the symlinks allows the jailkited users to access `/var/www/clients/clientXX/webYY/web` and `/var/www/clients/clientXX/webYY/website.git` using the full path.

Reply:
What @clemens-tolboom meant is that the $docroot van is concatenated twice

Reply:
Yes, and that is what I explained in the first reply

Reply:
Checking with http://php.net/manual/en/function.is-dir.php

```php
// Say
$docroot = '/var/www/clients/clientXX/webYY';

// $docroot . '/' . $docroot === '/var/www/clients/clientXX/webYY' '/' '/var/www/clients/clientXX/webYY' ?
```

Me puzzeld still.

Reply:
Yes, `$docroot . $docroot` === `'/var/www/clients/clientXX/webYY''/var/www/clients/clientXX/webYY'`.  
The thing is that users with a Jailkit chroot environment has the `$docroot` as their root directory (/). And since there's no `/var/www/clients/....` in the `$docroot` and the git hook uses the environment of the SSH user there would be no `/var/www/clients/....` directory to access.  
Basically it is just a wonky workaround so that both users with a jailkit and NONE chroot env to be able to access /var/www/clients/clientXX/webYY.  
  
So, from the perspective of a Jailkit chroot user the `/var/www/clients/clientXX/webYY/var/www/clients/clientXX/webYY` directory would only be `/var/www/clients/clientXX/webYY` since their root (`/`) is `/var/www/clients/clientXX/webYY`
To try to make it more clear, here is the output of "pwd" from a jailkit vs NONE chroot user
```
web30@cp:~$ # NONE chroot Environment
web30@cp:~$ pwd
/var/www/clients/client1/web30/home/grid_servers_testuser
web30@cp:~$ ls -hal ../../
total 20K
total 84K
drwxr-xr-x 21 root                root    4.0K Jul 26  2017 .
drwxr-xr-x 21 root                root    4.0K Jul 26  2017 ..
-rwxr-x---  1 web30               client1    0 Jul  3 13:00 .bash_history
-rw-r--r--  1 web30               client1    0 Jul  3 13:00 .profile
drwx------  2 web30               client1 4.0K Jul  3 13:00 .ssh
drwxr-xr-x  2 root                root    4.0K Feb 22 02:24 bin
drwxr-xr-x  2 web30               client1 4.0K Jul 26  2017 cgi-bin
drwxr-xr-x  2 root                root    4.0K Nov  9  2017 dev
drwxr-xr-x  7 root                root    4.0K Jun 29 01:24 etc
drwx--x--x  2 web30               client1 4.0K Jul 26  2017 git
drwxr-xr-x  5 root                root    4.0K Jul  3 12:33 home
drwxr-xr-x  4 root                root    4.0K Jul 26  2017 lib
drwxr-xr-x  2 root                root    4.0K Jul 26  2017 lib64
drwxr-xr-x  2 root                root    4.0K Jul  4 22:02 log
drwx--x---  2 web30               client1 4.0K Jul 26  2017 private
drwxr-xr-x  3 root                root    4.0K Jul 26  2017 run
drwxr-xr-x  2 root                root    4.0K Jul 26  2017 ssl
drwxrwxrwx  2 web30               client1 4.0K Jul 26  2017 tmp
drwxr-xr-x  6 root                root    4.0K Jul 26  2017 usr
drwxr-xr-x  4 root                root    4.0K Jul  5 12:34 var
drwx--x--x  4 web30               client1 4.0K Jul  5 12:36 web
drwx--x---  2 web30               client1 4.0K Jul 26  2017 webdav
drwxr-xr-x  8 web30               client1 4.0K Jul  5 12:36 website.git
```  

```
grid_servers_rynoxx@domainname.com:~$ # jailkit chroot Environment
grid_servers_rynoxx@domainname.com:~$ pwd
/home/grid_servers_rynoxx
grid_servers_rynoxx@domainname.com:~$ ls -hal ../../
total 84K
drwxr-xr-x 21 root                root    4.0K Jul 26  2017 .
drwxr-xr-x 21 root                root    4.0K Jul 26  2017 ..
-rwxr-x---  1 grid_servers_rynoxx client1    0 Jul  3 13:00 .bash_history
-rw-r--r--  1 grid_servers_rynoxx client1    0 Jul  3 13:00 .profile
drwx------  2 grid_servers_rynoxx client1 4.0K Jul  3 13:00 .ssh
drwxr-xr-x  2 root                root    4.0K Feb 22 02:24 bin
drwxr-xr-x  2 grid_servers_rynoxx client1 4.0K Jul 26  2017 cgi-bin
drwxr-xr-x  2 root                root    4.0K Nov  9  2017 dev
drwxr-xr-x  7 root                root    4.0K Jun 29 01:24 etc
drwx--x--x  2 grid_servers_rynoxx client1 4.0K Jul 26  2017 git
drwxr-xr-x  5 root                root    4.0K Jul  3 12:33 home
drwxr-xr-x  4 root                root    4.0K Jul 26  2017 lib
drwxr-xr-x  2 root                root    4.0K Jul 26  2017 lib64
drwxr-xr-x  2 root                root    4.0K Jul  4 22:02 log
drwx--x---  2 grid_servers_rynoxx client1 4.0K Jul 26  2017 private
drwxr-xr-x  3 root                root    4.0K Jul 26  2017 run
drwxr-xr-x  2 root                root    4.0K Jul 26  2017 ssl
drwxrwxrwx  2 grid_servers_rynoxx client1 4.0K Jul 26  2017 tmp
drwxr-xr-x  6 root                root    4.0K Jul 26  2017 usr
drwxr-xr-x  4 root                root    4.0K Jul  5 12:34 var
drwx--x--x  4 grid_servers_rynoxx client1 4.0K Jul  5 12:36 web
drwx--x---  2 grid_servers_rynoxx client1 4.0K Jul 26  2017 webdav
drwxr-xr-x  8 grid_servers_rynoxx client1 4.0K Jul  5 12:36 website.git
```  

Both of those two are from the same ISPConfig website (client1, web30), the first has no chroot env while the second is using jailkit.  
So, while the first one has no problem accessing `/var/clients/client1/web30/`, when the second one tries to access `/var/clients/client1/web30/` it is in reality trying to access `/var/clients/client1/web30/var/clients/client1/web30/`  


I hope this clears things up for you! :smile:

Reply:
Thanks for the elaborated answer. Case closed. Sorry for waisting us time :-p